File: 290-word-pattern/word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        
        if(pattern.length() != words.length) {
            return false;
        }

        HashMap<Character, String>charToWord = new HashMap<>();
        HashMap<String, Character>wordToChar = new HashMap<>();

        for(int i = 0; i < words.length; i++) {
            char charP = pattern.charAt(i);
            String word = words[i];

            if(charToWord.containsKey(charP)) {
                if(!charToWord.get(charP).equals(word)) {
                    return false;
                }
            } else {
                charToWord.put(charP, word);
            }

            if(wordToChar.containsKey(word)) {
                if (wordToChar.get(word)!=charP) {
                    return false;
                }
            } else {
                wordToChar.put(word,charP);
            }
        
        }
        return true;
    }
}

File: 290-word-pattern/word-pattern.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        words = s.split()

        if len(pattern) != len(words):
            return False
        
        char_to_word = {}
        word_to_char = {}
    
        for a,b in zip(pattern, words):
            if a in char_to_word:
                if char_to_word[a] !=b:
                    return False
            else: 
                char_to_word[a] = b

            if b in word_to_char:
                if word_to_char[b] != a:
                    return False
            else:
                word_to_char[b] = a
        
        return True